pep-0695.rst

```diff
@@ -0,0 +1,1103 @@
+in scope for function decorators.
+
+::
+
+    def func1[T](a: T) -> T: ...  # OK
+
+    print(T)  # Runtime error: 'T' is not defined
+
+    def func2[T](a = list[T]): ...  # Runtime error: 'T' is not defined
+
+    @dec(list[T])  # Runtime error: 'T' is not defined
+    def func3[T](): ...
+
+
+Upper Bound Specification
+-------------------------
+
+For a non-variadic type parameter, an "upper bound" type can be specified
+through the use of a type annotation expression. If an upper bound is
+not specified, the upper bound is assumed to be ``object``.
+
+::
+
+    class ClassA[T: str]: ...
+
+The specified upper bound type must use an expression form that is allowed in
+type annotations. More complex expression forms should be flagged
+as an error by a type checker. Quoted forward declarations are allowed.
+
+The specified upper bound type must be concrete. An attempt to use a generic
+type should be flagged as an error by a type checker.
+
+::
+
+    class ClassA[T: dict[str, int]]: ...  # OK
+
+    class ClassB[T: "ForwardDeclaration"]: ...  # OK
+
+    class ClassC[T: dict[str, V]]: ...  # Type checker error: generic type
+
+    class ClassD[T: [str, int]]: ...  # Type checker error: illegal expression form
+
+
+Constrained Type Specification
+------------------------------
+
+For a non-variadic type parameter, a set of two or more "constrained types"
+can be specified through the use of a literal tuple expression that contains
+two or more types.
+
+::
+
+    class ClassA[AnyStr: (str, bytes)]: ...  # OK
+
+    class ClassB[T: ("ForwardDeclaration", bytes)]: ...  # OK
+
+    class ClassC[T: ()]: ...  # Type checker error: two or more types required
+
+    class ClassD[T: (str, )]: ...  # Type checker error: two or more types required
+
+    t1 = (bytes, str)
+    class ClassE[T: t1]: ...  # Type checker error: literal tuple expression required
+
+
+If the specified type is not a tuple expression or the tuple expression includes
+complex expression forms that are not allowed in a type annotation, a type
+checker should generate an error. Quoted forward declarations are allowed.
+
+::
+
+    class ClassF[T: (3, bytes)]: ...  # Type checker error: invalid expression form
+
+
+The specified constrained types must be concrete. An attempt to use a generic
+type should be flagged as an error by a type checker.
+
+::
+
+    class ClassG[T: (list[S], str)]: ...  # Type checker error: generic type
+
+
+
+Generic Type Alias
+------------------
+
+We propose to introduce a new statement for declaring type aliases. Similar
+to ``class`` and ``def`` statements, a ``type`` statement defines a scope
+for type parameters. 
+
+::
+
+    # A non-generic type alias
+    type IntOrStr = int | str
+
+    # A generic type alias
+    type ListOrSet[T] = list[T] | set[T]
+
+
+Type aliases can refer to themselves without the use of quotes.
+
+::
+
+    # A type alias that refers to a forward-declared type
+    type AnimalOrVegetable = "Animal" | "Vegetable"
```

Comment:
Note that if the RHS is executed, this is a runtime error because str doesn't define `__or__`.